fix requireNoNulls(), filterNotNull(), withIndex() subscription leaks

dsvoronin opened the pull request:
address #38

dsvoronin committed:
fix requireNoNulls(), filterNotNull(), withIndex() subscription leaks

zsxwing commented on src/main/kotlin/rx/lang/kotlin/observables.kt:
```diff
@@ -58,34 +58,20 @@ public fun <T, R> Observable<T>.lift(operator : (Subscriber<in R>) -> Subscriber
  */
-public fun <T> Observable<T>.withIndex() : Observable<IndexedValue<T>> = lift { s ->
+public fun <T> Observable<T>.withIndex() : Observable<IndexedValue<T>> {
     var index = 0
```
This is incorrect. The returned `Observable` may be subscribed with multiple `Subscriber`s.

dsvoronin replied:
You speaking of thread safe access to index, right?

zsxwing replied:
Not. E.g.,

```
o2 = o1.withIndex()
o2.subscribe(s1)
o2.subscribe(s2)
```

Each `subscribe` should use a new `index`. But in your current implementation, it shares the same `index` between different `subscribe`s.

I know nothing about Kotlin. But in RxJava, you can implement it using `o.zipWith(Observable.range(0, Integer.MAX_VALUE))`.

dsvoronin replied:
ok, i'll add test for multiple subscribers and fix it somehow. thanks for review!

cy6erGn0m commented on src/main/kotlin/rx/lang/kotlin/observables.kt:
```diff
@@ -58,34 +58,20 @@ public fun <T, R> Observable<T>.lift(operator : (Subscriber<in R>) -> Subscriber
-            onError { t -> s.onError(t) }.
-            onNext { v -> if (v != null) s.onNext(v) }
-}
+public fun <T : Any> Observable<T?>.filterNotNull(): Observable<T> = filter { it != null }.map { it as T }
```
what is the idea of `map { it as T }` ? As far as I remember it will produce `Observable<T!>` rather than `Observable<T>`. Did you tried it?

dsvoronin replied:
filter{it!=null} -> (we still have T?, but T in runtime because no null values will be passed) (compiler can't smart cast here, because kotlin is not aware of rx filter), so we need somehow  cast value ->
map {it as T} will give us T! , you're right, but it compiles correct, and we won't get runtime exceptions.

Some kind of hack i know. Do you have better solution?

dsvoronin committed:
fix withIndex for multiple subscribers
possibly better way to handle fllterNotNull()

dsvoronin commented:
@zsxwing i added test for multiple subscribers, and used range as you suggested
@cy6erGn0m i removed rx map operator, and just unsafe casted it to Observable<T>, it still works correctly, but maybe less hacky now. What do you think?

MarioAriasC commented:
Is this ready to merge?

cy6erGn0m commented:
@MarioAriasC looks good to me

MarioAriasC commented:
ok
